File: src/main/java/pl/allegro/experiments/chi/chiserver/infrastructure/druid/DruidClient.java
package pl.allegro.experiments.chi.chiserver.infrastructure.druid;

import org.springframework.web.client.RestTemplate;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;

public class DruidClient {
    private final String druidApiHost;
    private final RestTemplate restTemplate;

    public DruidClient(String druidApiHost, RestTemplate restTemplate) {
        this.druidApiHost = druidApiHost;
        this.restTemplate = restTemplate;
    }

    public String query(String body) {
        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON_UTF8);
        HttpEntity<String> query = new HttpEntity<>(body, headers);
        try {
            return restTemplate.postForEntity(
                    "http://" + druidApiHost + "/druid/v2/?pretty",
                    query, String.class
            ).getBody();
        } catch (Exception e) {
            throw new DruidException("Error while trying to get data from " + druidApiHost, e);
        }
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/infrastructure/druid/DruidClient.kt
package pl.allegro.experiments.chi.chiserver.infrastructure.druid

import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType
import org.springframework.web.client.RestTemplate
import java.time.LocalDate
import java.time.format.DateTimeFormatter

/**
 * Converts given date into string representing Druid intervals for one day.
 */
fun oneDayIntervals(date: LocalDate): String {
    val dateStr = DateTimeFormatter.ISO_LOCAL_DATE.format(date)
    return "${dateStr}T00Z/${dateStr}T23:59:59.999Z"
}

/**
 * Returns Druid intervals for last day metrics.
 */
fun lastDayIntervals() = oneDayIntervals(LocalDate.now().minusDays(1))
